purchase_stock_bl_date/models/stock_picking.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    bl_date = fields.Date()
+    force_bl_date = fields.Boolean()
+
+    def action_toggle_is_locked(self):
+        if not self.is_locked and self.state == "done" and self.force_bl_date:
+            for move in self.move_ids_without_package:
+                move.write(
+                    {
+                        "price_unit": move._get_price_unit(),
+                    }
+                )
+                if move.product_id.cost_method == "average":
+                    move.cost_price_avco_sync()
+                else:
+                    move._run_valuation()
+                move.update_amount_journal_entry()
```

Comment:
Am I correct to understand that this simply overrides the value of the receipt stock move and its journal entry?  In that case, it adds the risk of inconsistent balances in case the inventory has already been shipped out (in FIFO context).  Is there a way to prevent that from happening (I suppose the same applies to AVCO - has this issue/concern been addressed somewhere)?

Reply:
Yes, you right. I have no idea how to prevent in case the inventory has already been shipped out at the moment.

Reply:
@kanda999 @rinaldifirdaus  Please let me know if you guys have the advice for this case. Current situation for recovery can recalculate the journal entry and stock valuation.

Reply:
in manual operation, to do recovery of the inventory that has been shipped out, we can adjust the product's done qty on the delivery record to be 0 (same with the logic on receipt). when we set the delivered qty to 0, valuation of the product will be back to as it is on inventory valuation layer regardless of the current applied unit price.
then the flow would be:
1. manually set the delivered qty to 0 (valuation will be corrected with amount when it was delivered)
2. set the force bl date as true, update/revise the bl date 
3. manually set back the delivered qty to its original qty on the delivery record (valuation will be corrected with the new amount based on bl date correction)

i think we may need to ask the client to recover the delivery manually (without customization) then update the bl date as above flow. 
or maybe you have an idea to automate above flow @AungKoKoLin1997 ?
it might be tricky to automate set the delivered qty to 0 since on the delivery record may have various products (not only the inventory that we want to recover) on the SO line. but it might be ok to temporarily set all SO line qty to 0 and set back all the SO line qty as it was after update/revise the bill date (need to check).

Reply:
>  maybe you have an idea to automate above flow

It is very hard for me to implement this and this is very complex.